.github/workflows/publish.yml

```diff
@@ -14,38 +17,99 @@ on:
 # or [0-9]+.[0-9]+.[0-9]+-rc optionally followed by digits
 
 jobs:
-  validate-tag:
-    runs-on: ubuntu-latest
+  # If we use the viamrobotics/build-action@v1, it farms out compilation to Linux machines no
+  # matter which OS you're targeting (which is fine for Go and Python, but not for C++).
+  # Furthermore, these Linux machines do not cache build artifacts from one run to another. So
+  # instead, we're going to build everything on Github's action runners directly.
+  build-windows:
+    name: Build for Windows
+    runs-on: windows-2019
     steps:
-      - name: Validate tag format
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Restore cached libraries
+        uses: actions/cache@v3
+        with:
+          path: C:\Users\runneradmin\.conan2
+          key: windows-2019-conan-${{ hashFiles('**/conanfile.py') }}
+          restore-keys: |
+            windows-2019-conan-
+      - name: Setup and build
         run: |
-          TAG="${{ github.event.release.tag_name }}"
-          echo "Validating tag: $TAG"
-          if [[ $TAG =~ ^[0-9]+\.[0-9]+\.[0-9]+(-rc[0-9]*)?$ ]]; then
-            echo "Tag matches semver."
-          else
-            echo "Error: tag does not match semver"
-            exit 1
-          fi
-  publish:
-    needs: validate-tag
-    name: Build and Upload module
-    runs-on: ubuntu-latest
+          make setup
+          make module.tar.gz
+      # As of April 2025, there isn't an obvious way to upload to the Viam module registry from
+      # Windows. So, we upload to the Github Actions artifacts, then download it on Linux in a later
+      # step to upload to the module registry.
+      - name: Store the module as a temporary artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: windows-2019-module-${{ github.sha }}
+          path: ./module.tar.gz
+
+  build-unix:
+    name: Build for Unix
+    strategy:
+      matrix:
+        build_target:
+          - runs_on: ubuntu-latest
+          - runs_on: ubuntu-24.04-arm
+          - runs_on: macos-latest
+          - runs_on: macos-13
+    runs-on: ${{ matrix.build_target.runs_on }}
     steps:
       - name: Checkout
         uses: actions/checkout@v4
-
-      - uses: actions/cache@v3
+      - name: Restore cached libraries
+        uses: actions/cache@v3
         with:
           path: ~/.conan2
-          key: ${{runner.os}}-conan-${{ hashFiles('**/conanfile.py') }}
+          key: ${{ matrix.build_target.runs_on }}-conan-${{ hashFiles('**/conanfile.py') }}
```

Comment:
Sneaky trick alert: if the cache key is `runner.os`, the Ubuntu-arm and Ubuntu-amd64 runners share a cache, and you get errors that you're trying to use amd64-only command-line arguments in the arm compiler or vice versa. So, the cache key needs to be based on both the OS _and_ the CPU architecture. I'm not going to put a comment in the code because when this PR is merged it will look obvious and no one will think of changing it.